airflow/creds/aws_ssm.py

```diff
@@ -0,0 +1,127 @@
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""
+Objects relating to sourcing connections from AWS SSM Parameter Store
+"""
+
+from typing import List
+
+import boto3
+
+from airflow import conf
+from airflow.creds import CONN_ENV_PREFIX, BaseCredsBackend
+from airflow.models import Connection
+
+
+class AwsSsmCredsBackend(BaseCredsBackend):
+    """
+    Retrieves Connection object from AWS SSM Parameter Store
+
+    Configurable via airflow config like so:
+
+    .. code-block:: ini
+
+        [aws_ssm_creds]
+        ssm_prefix = /airflow
+        profile_name = default
+
+    For example, if ssm path is ``/airflow/AIRFLOW_CONN_SMTP_DEFAULT``, this would be accessible if you
+    provide ``ssm_prefix = /airflow`` and conn_id ``smtp_default``.
+
+    """
+
+    conf_section = "aws_ssm_creds"
+    conf_key_ssm_prefix = "ssm_prefix"
+    default_prefix = "/airflow"
+    conf_key_profile_name = "profile_name"
+    default_profile = "default"
+
+    def __init__(self, *args, **kwargs):
+        pass
+
+    @property
+    def ssm_prefix(self) -> str:
+        """
+        Gets ssm prefix from conf.
+
+        Ensures that there is no trailing slash.
+
+        :return:
+        """
+        ssm_prefix = conf.get(
+            section=self.conf_section,
+            key=self.conf_key_ssm_prefix,
+            fallback=self.default_prefix,
+        )
+        if ssm_prefix[-1] == "/":
+            ssm_prefix = ssm_prefix[:-1]
+        return ssm_prefix
+
+    @property
+    def aws_profile_name(self) -> str:
+        """
+        Gets AWS profile to use from conf.
+
+        :return:
+        """
+        profile_name = conf.get(
+            section=self.conf_section,
+            key=self.conf_key_profile_name,
+            fallback=self.default_profile,
+        )
+        return profile_name
+
+    def build_ssm_path(self, conn_id: str):
+        """
+        Given conn_id, build SSM path.
+
+        Assumes connection params use same naming convention as env vars, but may have arbitrary prefix.
+
+        :param conn_id:
+        :return:
+        """
+        param_name = (CONN_ENV_PREFIX + conn_id).upper()
+        param_path = self.ssm_prefix + "/" + param_name
+        return param_path
+
+    def get_conn_uri(self, conn_id):
+        """
+        Get param value
+
+        :param conn_id:
+        :return:
+        """
+        session = boto3.Session(profile_name=self.aws_profile_name)
```

Comment:
Does this work fine with instance profiles with an empty `.aws/config`?

Reply:
this is a good point.  if there is no `.aws/config`, then this would fail.  the solution is to use `None` as fallback default for `self.aws_profile_name` instead of `'default'`.  then if aws creds are available in env params, boto3 will use them without error.